### ISSUE TITLE
SVG resolution

### ISSUE THREAD

**Issue by jimthedev** [andrieslouw/imagesweserv#106]
Hi there, great service. I have a few svgs that are becoming pixelated when going through the service. Not a huge deal but does not seem to be related to any output or quality settings that I can configure.

Original (expected):
http://slack.mnphp.org/badge.svg

Tranformed (grainy):
https://images.weserv.nl/?url=slack.mnphp.org/badge.svg

Any chance you know if this something that will be fixed when #94 is merged?

**Comment by andrieslouw** [comment]
Will be fixed by #94 indeed, our staging environment shows way better results (uses latest development build of libvips), although it ain't perfect either. Good test case, thanks for providing it! (We're still experimenting with good SVG support).

**Comment by jimthedev** [comment]
Great, thanks for the update. Keep up the good work. Leaving open for posterity until #94 merges. Cheers!

**Comment by kleisauke** [comment]
Fix has been rolled out to production. The cause of this bug was due to missing fonts on the production server.

Cache has been invalidated for the [particular image](https://images.weserv.nl/?url=slack.mnphp.org/badge.svg). If you want, you can send us an [email](mailto:admin@weserv.nl) with the images that are affected by this bug and need to be purged from cache.

Let me know if this looks better.

**Comment by kleisauke** [comment]
Update:
SVG images without `width` and `height` attributes are still pixelated when being processed. This specific issue comes from a regression in libvips v8.5.7+. See https://github.com/jcupitt/libvips/pull/724.

Hopefully it will be solved in libvips 8.6 (due in a few weeks). I'll update this issue once I have more information.

**Comment by jimthedev** [comment]
Thanks for the update @kleisauke. Glad that use-case is being included in the mix.

**Comment by kleisauke** [comment]
Update:
Today we updated the production server from CentOS 7.3.1611 to CentOS 7.4.1708. With this update `librsvg` has also updated (from 2.39.0-1 to 2.40.16-1).

Unfortunately, `librsvg` has now some problems with provided SVG example, see:
![badge](https://user-images.githubusercontent.com/12746591/30512567-b5d9944c-9af2-11e7-84b4-f09dc050f881.png)

This is fixable if we remove the unnecessary rectangle from the SVG (the id `g` doesn't exits in the SVG).
```diff
--- badge.svg	Sat Sep 16 16:40:15 2017
+++ badge.svg	Sat Sep 16 16:48:45 2017
@@ -2,7 +2,6 @@
   <rect rx="3" width="94" height="20" fill="#555"></rect>
   <rect rx="3" x="47" width="47" height="20" fill="#E01563"></rect>
   <path d="M47 0h4v20h-4z" fill="#E01563"></path>
-  <rect rx="3" width="94" height="20" fill="url(#g)"></rect>
   <g text-anchor="middle" font-family="Verdana" font-size="11">
     <text fill="#010101" fill-opacity=".3" x="24" y="15">slack</text>
     <text fill="#fff" x="24" y="14">slack</text>
```

With the above patch, the processed SVG looks like this:
![badge](https://user-images.githubusercontent.com/12746591/30512595-672b21e8-9af3-11e7-80ea-ab6481e4185f.png)

The [pixelated problem](https://images.weserv.nl/?url=t0.nl/check.svg&w=1024&t=fitup) isn't fixed yet, therefore https://github.com/jcupitt/libvips/pull/724 needs to be merged first. In the meantime, I've patched libvips at our test environment. If you want, you can test it here:
https://t0.nl/?url=t0.nl/check.svg&w=1024&t=fitup

(I can't recommend this for production images, but it will be fine for one time processing)

**Comment by jimthedev** [comment]
Hah, the extra g actually broke Safari too and was fixed a while ago but this organization must be using an old version of Slackin

https://github.com/rauchg/slackin/commit/edcaebce0fb2b4e07ca834a58fa069c53a5e0bf3#diff-09a4fe039e6455db71dfaa055aec0b53

**Comment by kleisauke** [comment]
The [pixelated problem](https://images.weserv.nl/?url=t0.nl/check.svg&w=1024&t=fitup) has been fixed. A unit test for this specific issue has been added, to ensure that this does not happen again.

Thanks for your patience, don't hesitate to make a new issue if you find something else that needs to be fixed.

**Comment by muddman** [comment]
Thanks for providing the service.

Unfortunately I'm having a similar issue with any badge i try.

| Source | Image Style | Original  | Weserv |
| ------------- | ------------- | ------------- | ------------- |
| shield.io | plastic | ![shields.io svg](https://img.shields.io/github/v/release/clearasmudd/windows-formula?style=plastic)  | ![ver](https://images.weserv.nl/?url=//img.shields.io/github/v/release/clearasmudd/windows-formula?style=plastic) |
| shield.io | flat | ![shields.io svg](https://img.shields.io/github/v/release/clearasmudd/windows-formula?style=flat)  | ![ver](https://images.weserv.nl/?url=//img.shields.io/github/v/release/clearasmudd/windows-formula?style=flat) |
| shield.io | flat-square| ![shields.io svg](https://img.shields.io/github/v/release/clearasmudd/windows-formula?style=flat-square)  | ![ver](https://images.weserv.nl/?url=//img.shields.io/github/v/release/clearasmudd/windows-formula?style=flat-square) |
| shield.io | for-the-badge | ![shields.io svg](https://img.shields.io/github/v/release/clearasmudd/windows-formula?style=for-the-badge)  | ![ver](https://images.weserv.nl/?url=//img.shields.io/github/v/release/clearasmudd/windows-formula?style=for-the-badge) |
| shield.io | social | ![shields.io svg](https://img.shields.io/github/v/release/clearasmudd/windows-formula?style=social)  | ![ver](https://images.weserv.nl/?url=//img.shields.io/github/v/release/clearasmudd/windows-formula?style=social) |
| requires.io |  | ![shields.io svg](https://requires.io/github/clearasmudd/windows-formula/requirements.svg?branch=master)  | ![ver](https://images.weserv.nl/?url=//requires.io/github/clearasmudd/windows-formula/requirements.svg?branch=master) |
| readthedocs  | | ![](https://readthedocs.com/projects/clear-as-mudd-windows-formula/badge/?version=master)  | ![ver](https://images.weserv.nl/?url=//readthedocs.com/projects/clear-as-mudd-windows-formula/badge/?version=master) |
| codacity | | ![](https://api.codacy.com/project/badge/Grade/dbf7bbe2183b4dfe8e1e8736ee48718c) |  ![ver](https://images.weserv.nl/?url=//api.codacy.com/project/badge/Grade/dbf7bbe2183b4dfe8e1e8736ee48718c)|
| snyk.io | | ![](https://snyk.io/test/github/clearasmudd/windows-formula/badge.svg) |  ![ver](https://images.weserv.nl/?url=//snyk.io/test/github/clearasmudd/windows-formula/badge.svg)|

**Comment by kleisauke** [comment]
@muddman It looks OK here:
![screenshot](https://user-images.githubusercontent.com/12746591/75478035-88ea1080-599d-11ea-83c6-5166d96079b8.png)

There's only a `<` missing within the `shield.io social` badge and the `shield.io for-the-badge` text looks odd, I'll investigate that.

Perhaps your zoom level within the browser is not 100% (i.e. 1:1 ratio)? With 110% zoom on Google Chrome I see this:
![screenshot-110-zoom](https://user-images.githubusercontent.com/12746591/75478178-cb135200-599d-11ea-8444-02325984187b.png)

**Comment by muddman** [comment]
Thanks.

I confirmed the browser zoom level is 100% on Windows 10 1909 Build 18363.657

**This shows the zoom level**
![image](https://user-images.githubusercontent.com/4186835/75482279-7f977e80-5972-11ea-881f-0387548d7892.png)

**Screenshot in Chrome 80.0.3987.122**

![image](https://user-images.githubusercontent.com/4186835/75481458-12371e00-5971-11ea-99b5-15fe4c859a4f.png)

**Screenshot in Edge**

![image](https://user-images.githubusercontent.com/4186835/75481630-5a564080-5971-11ea-9dc1-a550e3784031.png)

**Screenshot from an iPad Pro (10.5-inch) , iOS 13.3.1**

![SVG resolution  Issue #106  weservimages  GitHub](https://user-images.githubusercontent.com/4186835/75482745-63481180-5973-11ea-8ccc-7568b8571df0.png)

**Comment by andrieslouw** [comment]
I'm unsure if it is possible to "fix" this, good [font rasterization](https://en.wikipedia.org/wiki/Font_rasterization) is dependent the screentype, resolution, up- or downscaling, etc. And while these are things which a browser will be able to detect and handle when using SVG's, we are unable to detect and/or incorporate them to make pixel perfect versions for every environment (eg. LCD is different from OLED). In addition; many of the rasterization techniques used by Microsoft and Apple are patented and/or closed-source.

For example, this (original SVG):
https://upload.wikimedia.org/wikipedia/commons/4/4e/Meta_SVG_fonts.svg

Will look different from this (SVG rasterized by us):
https://images.weserv.nl/?url=upload.wikimedia.org/wikipedia/commons/4/4e/Meta_SVG_fonts.svg&bg=fff

But our rasterization matches this (SVG rasterized by Wikimedia):
https://upload.wikimedia.org/wikipedia/commons/thumb/4/4e/Meta_SVG_fonts.svg/1024px-Meta_SVG_fonts.svg.png

Please let us know if there is a SVG library around which is able to do a better job of font rasterization to bitmaps.

**Comment by muddman** [comment]
makes sense -- thanks!
